src/engine/renderer/glsl_source/cull_cp.glsl

```diff
@@ -73,21 +75,97 @@ struct Plane {
+
+	float vy = sqrt( center.y * center.y + czr2 );
+	float miny = center.y * center.y + czr2 >= 0.0 ? ( vy * center.y - cr.z ) / ( vy * center.z + cr.y ) : -1.0;
+	float maxy = center.y * center.y + czr2 >= 0.0 ? ( vy * center.y + cr.z ) / ( vy * center.z - cr.y ) : 1.0;
+
+	boundingBox = vec4( minx * P00, miny * P11, maxx * P00, maxy * P11 );
+	boundingBox = boundingBox.xwzy * vec4( 0.5f, -0.5f, 0.5f, -0.5f ) + vec4( 0.5, 0.5, 0.5, 0.5 ); // clip space -> uv space
+
+	return true;
+}
 
 bool CullSurface( in BoundingSphere boundingSphere ) {
+    bool culled = false;
+
+    // Frustum culling
     // Skip far plane because we always make it encompass the whole map in the current direction
     // This might need to be changed later for shadowmaps since lights could have some far plane set
     for( int i = 0; i < 5; i++ ) {
         const float distance = dot( u_Frustum[i].xyz, boundingSphere.center ) - u_Frustum[i].w;
 
         if( distance < -boundingSphere.radius ) {
-            return true;
+            culled = u_UseFrustumCulling;
         }
     }
-    return false;
+    
+    // Occlusion culling for surfaces that passed frustum culling
+    vec4 boundingBox = vec4( -1.0, 1.0, -1.0, 1.0 );
+    const vec3 viewSpaceCenter = vec3( u_ModelViewMatrix * vec4( boundingSphere.center, 1.0 ) );
+    
+    // Only do occlusion culling on the main view because portals use stencil buffer,
+    // so doing depth reduction there would be complicated
+
+    // ProjectSphere() gives us the screen-space AABB of the surface, based on its bounding sphere
+    if( u_UseOcclusionCulling && u_ViewID == 0 && !culled
+        && ProjectSphere( viewSpaceCenter, boundingSphere.radius, r_zNear, u_P00, u_P11, boundingBox ) ) {
+        boundingBox.xz = vec2( min( 1 - boundingBox.x, 1 - boundingBox.z ), max( 1 - boundingBox.x, 1 - boundingBox.z ) );
+        const float width = clamp( ( boundingBox.z - boundingBox.x ) * float( u_ViewWidth ), 0.0, float( u_ViewWidth ) );
+		const float height = clamp( ( boundingBox.w - boundingBox.y ) * float( u_ViewHeight ), 0.0, float( u_ViewHeight ) );
+
+        // Sample the depth-reduction image at the mip-level where the surface covers 4 pixels
+        const int level = int( floor( log2( max( width, height ) ) ) );
+
+        // Coords for the 4 pixels covered by the AABB, adjusted for partially off-screen surfaces as needed
+        vec2 surfaceCoords = vec2( u_ViewWidth >> level, u_ViewHeight >> level );
+        surfaceCoords *= ( boundingBox.xy + boundingBox.zw ) * 0.5;
+        const ivec2 surfaceCoordsFloor = ivec2( clamp( surfaceCoords.x, 0.0, float( u_ViewWidth >> level ) ), clamp( surfaceCoords.y, 0.0, float( u_ViewHeight >> level ) ) );
+        ivec4 depthCoords = ivec4( surfaceCoordsFloor.xy,
+                                   surfaceCoordsFloor.x + ( surfaceCoords.x - surfaceCoordsFloor.x >= 0.5 ? 1 : -1 ),
```

Comment:
The object won't always occupy 2 pixels in both dimensions, right? It could have some oblong shape and be 1 pixel wide in the x axis and 2 in the y axis for example. I understand this is not a bug as the worst thing that could happen here is a surface not being culled, but I wonder if there isn't another way to do the calculation.

Reply:
That can happen sometimes, yes. I've tried wrapping the texel fetches in if statements so only the first one is guaranteed and the next ones are only done if the new coordinates don't match any of the previous ones, but some surfaces became invisible when they shouldn't be. Not sure why as I couldn't spot what the issue was. This can be improved later though.

Reply:
There's also an extension that would allow doing this with just one regular texture sample, but the availability of that extension is not enough to be the only option.

Reply:
Ah wait, it wasn't working properly because I accidentally used `min` instead of `max`. Anyway though, with that change it seems to be a tiny bit slower (<0.1ms). Any consecutive texel fetches with the same coordinates will likely just come from the same texture cache anyway.